flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -189,66 +179,62 @@ protected void register(String contenderID, LeaderContender contender) throws Ex
         checkNotNull(contender, "Contender must not be null.");
 
         synchronized (lock) {
-            Preconditions.checkState(
-                    leaderContender == null,
-                    "Only one LeaderContender is allowed to be registered to this service.");
-            Preconditions.checkState(
-                    this.contenderID == null, "The contenderID is only allowed to be set once.");
             Preconditions.checkState(
                     running,
                     "The DefaultLeaderElectionService should have established a connection to the backend before it's started.");
 
-            leaderContender = contender;
-            this.contenderID = contenderID;
+            leaderContenderRegistry.put(contenderID, contender);
 
             LOG.info(
                     "LeaderContender {} has been registered for {}.",
-                    contender.getDescription(),
+                    contenderID,
                     leaderElectionDriver);
 
             if (issuedLeaderSessionID != null) {
                 // notifying the LeaderContender shouldn't happen in the contender's main thread
                 runInLeaderEventThread(
-                        () -> notifyLeaderContenderOfLeadership(issuedLeaderSessionID));
+                        () ->
+                                notifyLeaderContenderOfLeadership(
+                                        contenderID, issuedLeaderSessionID));
             }
         }
     }
 
     @Override
     protected final void remove(String contenderID) {
         synchronized (lock) {
-            if (this.contenderID == null) {
+            if (!leaderContenderRegistry.containsKey(contenderID)) {
                 LOG.debug(
                         "The stop procedure was called on an already stopped DefaultLeaderElectionService instance. No action necessary.");
```

Comment:
update message to include the contender and clarify the the LES may still be running?

Reply:
You're right - I will update the message. We're not really "stopping" here anymore. That's some log message that survived the refactoring and should be rephrased.